src/sentry/sentry_metrics/consumers/indexer/slicing_router.py

```diff
@@ -0,0 +1,83 @@
+from typing import MutableMapping, Optional
+
+from arroyo import Message, Topic
+from arroyo.backends.kafka import KafkaPayload
+from confluent_kafka import Producer
+from django.conf import settings
+
+from sentry.ingest.partitioning import (
+    Sliceable,
+    is_sliced,
+    map_logical_partition_to_slice,
+    map_org_id_to_logical_partition,
+)
+from sentry.sentry_metrics.consumers.indexer.routing_producer import MessageRoute, MessageRouter
+from sentry.utils import kafka_config
+
+
+class SlicingRouter(MessageRouter[KafkaPayload]):
+    def __init__(
+        self,
+        sliceable: Sliceable,
+        logical_output_topic: str,
+    ) -> None:
+        self.__sliceable = sliceable
+        self.__logical_output_topic = logical_output_topic
+        self.__slice_to_producer: MutableMapping[int, MessageRoute] = {}
+
+        if not is_sliced(self.__sliceable):
+            self.__slice_to_producer[0] = MessageRoute(
+                producer=Producer(
+                    kafka_config.get_kafka_producer_cluster_options(
+                        settings.KAFKA_TOPICS[self.__logical_output_topic]["cluster"]
+                    )
+                ),
+                topic=Topic(self.__logical_output_topic),
+            )
+            self.__slicing_enabled = False
+        else:
+            for (
+                current_sliceable,
+                current_slice_id,
+            ), config in settings.SLICED_KAFKA_BROKER_CONFIG.items():
+                self.__slice_to_producer[current_slice_id] = MessageRoute(
+                    producer=Producer(config),
+                    topic=Topic(
+                        settings.SLICED_KAFKA_TOPIC_MAP[(current_sliceable, current_slice_id)]
+                    ),
+                )
+            assert len(self.__slice_to_producer) == len(
+                settings.SENTRY_SLICING_CONFIG[sliceable].keys()
+            )
+            self.__slicing_enabled = True
+
+    def get_route_for_message(self, message: Message[KafkaPayload]) -> MessageRoute:
+        """
+        Get route for the message. If slicing is enabled, the message will be routed
```

Comment:
The function does not seem to check "if slicing is enabled", just the presence of the `org_id` in the header

Reply:
Fixed in a later commit. Now there is an explicit check for slicing enabled at https://github.com/getsentry/sentry/pull/40776/files#diff-8b2a0552f8e4c6e830654b7fbf34e7ce3f14fc649178d415d711701e453124a8R60-R61

Reply:
SlicingRouter does not work in unsliced mode now. Fixed the comment and the code.